clean cephFS omap details

yati1998 opened the pull request:
Remove the omap object for the given subvolume.

yati1998 commented:
```Error: command terminated with exit code 1
did not load config file, using default settings.
2024-02-06T10:08:53.016+0000 7fa11474ae00 -1 Errors while parsing config file!
2024-02-06T10:08:53.016+0000 7fa11474ae00 -1 can't open ceph.conf: (2) No such file or directory
unable to get monitor info from DNS SRV with service name: ceph-mon
2024-02-06T10:08:53.018+0000 7fa11474ae00 -1 failed for service _ceph-mon._tcp
2024-02-06T10:08:53.018+0000 7fa11474ae00 -1 monclient: get_monmap_and_config cannot identify monitors to contact
failed to fetch mon config (--no-mon-config to skip)```

yati1998 commented:
cc: @travisn @Madhu-1

yati1998 commented:
> > https://github.com/rook/kubectl-rook-ceph/actions/runs/7799941151/job/21272736952?pr=237#step:12:20
> > ```
> > did not load config file, using default settings.
> > 2024-02-06T10:08:53.016+0000 7fa11474ae00 -1 Errors while parsing config file!
> > 2024-02-06T10:08:53.016+0000 7fa11474ae00 -1 can't open ceph.conf: (2) No such file or directory
> > unable to get monitor info from DNS SRV with service name: ceph-mon
> > 2024-02-06T10:08:53.018+0000 7fa11474ae00 -1 failed for service _ceph-mon._tcp
> > 2024-02-06T10:08:53.018+0000 7fa11474ae00 -1 monclient: get_monmap_and_config cannot identify monitors to contact
> > failed to fetch mon config (--no-mon-config to skip)```
> > ```
> 
> The plugin may not understand that it needs to append the mon and cluster commands to the `rados` command. Looks like you need to add the `rados` command [here](https://github.com/rook/kubectl-rook-ceph/blob/8a410dfd141c6e120f9ea5e2250b7b85a2452728/pkg/exec/exec.go#L109-L115).

That works @travisn . Thanks.

subhamkrai commented on .github/workflows/go-test.yaml:
```diff
@@ -77,7 +77,7 @@ jobs:
       - name: Subvolume command
         run: |
-          set -ex
+          set -e
```
```suggestion
          set -e
```
?

subhamkrai commented on pkg/filesystem/subvolume.go:
```diff
@@ -178,12 +183,28 @@ func Delete(ctx context.Context, clientsets *k8sutil.Clientsets, OperatorNamespa
 }
 
+// DeleteOmap deletes omap object and key for the given subvolume.
+func DeleteOmap(ctx context.Context, clientsets *k8sutil.Clientsets, OperatorNamespace, CephClusterNamespace, subVol, fs string) {
```
```suggestion
func deleteOmap(ctx context.Context, clientsets *k8sutil.Clientsets, OperatorNamespace, CephClusterNamespace, subVol, fs string) {
```

subhamkrai commented on pkg/filesystem/subvolume.go:
```diff
@@ -178,12 +183,28 @@ func Delete(ctx context.Context, clientsets *k8sutil.Clientsets, OperatorNamespa
 				continue
 			}
 			logging.Info("subvolume %q deleted", subvolume)
+			DeleteOmap(ctx, clientsets, OperatorNamespace, CephClusterNamespace, subvolume, fs)
```
```suggestion
			deleteOmap(ctx, clientsets, OperatorNamespace, CephClusterNamespace, subvolume, fs)
```

subhamkrai commented on pkg/filesystem/subvolume.go:
```diff
@@ -178,12 +183,28 @@ func Delete(ctx context.Context, clientsets *k8sutil.Clientsets, OperatorNamespa
+func DeleteOmap(ctx context.Context, clientsets *k8sutil.Clientsets, OperatorNamespace, CephClusterNamespace, subVol, fs string) {
+	omapdetails, err := getOmapDetails(ctx, clientsets, OperatorNamespace, CephClusterNamespace, subVol, fs)
+	if err != nil {
+		logging.Fatal(err)
```
```suggestion
		logging.Fatal(err)
```

it's better to good error message along with error

subhamkrai reviewed (commented):
Also, please update commit/PR title

yati1998 commented:
> The CI is failing:
> 
> ```
> + kubectl rook-ceph --operator-namespace test-operator -n test-cluster subvolume delete myfs test-subvol group-a
> Info: subvolume "test-subvol" deleted
> Error: error getting omap keys myfs-metadata/test.subvolume: (2) No such file or directory
> . failed to run command. command terminated with exit code 1%!(EXTRA string=failed to list omapval)
> Error: failed to get omap details: "error getting omap keys myfs-metadata/test.subvolume: (2) No such file or directory\n. failed to run command. command terminated with exit code 1"
> ```

Yes, this error is expected. Updated the test to include error in output.

Madhu-1 commented on pkg/filesystem/subvolume.go:
```diff
@@ -44,6 +45,11 @@ const (
 	staleWithSnapshot = "stale-with-snapshot"
 )
 
+type OmapDetails struct {
```
why this need to be exported?

yati1998 replied:
not required, will update it.

Madhu-1 commented on pkg/filesystem/subvolume.go:
```diff
@@ -211,7 +217,73 @@ func Delete(ctx context.Context, clientsets *k8sutil.Clientsets, OperatorNamespa
 }
+
+// deleteOmap deletes omap object and key for the given subvolume.
+func deleteOmap(ctx context.Context, clientsets *k8sutil.Clientsets, OperatorNamespace, CephClusterNamespace, subVol, fs string) {
```
what will happen to the omap if the subvolumes are deleted but the omap are not deleted/cleaned up?

yati1998 replied:
this is the case when someone deletes the subvolume from the backend. Here we are deleting the stale subvolume and just after that removing it's entry from Omap. 
Here, we are deleting omap of only stale subvolumes, as that's the input we are asking from the user.  In case we  want to cleanup the omap to remove all the stale omaps whose subvolume and pv both are deleted , we need to have separate command, where it will list all the omaps and pv's and verify the stale omap.

yati1998 replied:
will open and link one.

Madhu-1 commented on pkg/filesystem/subvolume.go:
```diff
@@ -211,7 +217,73 @@ func Delete(ctx context.Context, clientsets *k8sutil.Clientsets, OperatorNamespa
+	return false
+}
+
+func getOmapDetails(ctx context.Context, clientsets *k8sutil.Clientsets, OperatorNamespace, CephClusterNamespace, subVol, fs string) (OmapDetails, error) {
```
add function comments exactly what are we doing on each helper function as there are string operations it is highly difficult to understand what we are doing?

Madhu-1 commented on pkg/filesystem/subvolume.go:
```diff
@@ -211,7 +217,73 @@ func Delete(ctx context.Context, clientsets *k8sutil.Clientsets, OperatorNamespa
+		omapval1 = strings.Replace(omapval1, "nfs", "csi", 1)
+	} else {
+		omapval1 = strings.Replace(omapval, "vol", "volume", 1)
+
```
extra line

Madhu-1 commented on pkg/filesystem/subvolume.go:
```diff
@@ -211,7 +217,73 @@ func Delete(ctx context.Context, clientsets *k8sutil.Clientsets, OperatorNamespa
+func getOmapDetails(ctx context.Context, clientsets *k8sutil.Clientsets, OperatorNamespace, CephClusterNamespace, subVol, fs string) (OmapDetails, error) {
+
+	var details OmapDetails
+	var omapval1, pv string
```
use meaningful variable names, IMO we should not use `omapval1`

Madhu-1 commented on pkg/filesystem/subvolume.go:
```diff
@@ -211,7 +217,73 @@ func Delete(ctx context.Context, clientsets *k8sutil.Clientsets, OperatorNamespa
+	if ls != "" {
+		volname := strings.Split(ls, "|")
+		if checknfs {
+			pv = "csi.volume." + volname[12] + volname[14] + volname[16]
```
do we have length check for `volname` to ensure we don't panic, can add comments for this kind of operations

Madhu-1 reviewed (changes requested):
i think we should start testing all these functionalities in the CI to ensure it works, Please add test cases for it

Madhu-1 commented on pkg/filesystem/subvolume.go:
```diff
@@ -211,7 +217,73 @@ func Delete(ctx context.Context, clientsets *k8sutil.Clientsets, OperatorNamespa
+}
+
+// checkStaleSubvolume checks if there are any stale subvolume to be deleted
+func checkStaleSubvolume(ctx context.Context, clientsets *k8sutil.Clientsets, OperatorNamespace, CephClusterNamespace, fs, subvolume, svg string, k8sSubvolume map[string]subVolumeInfo) bool {
```
dont we already have a helper for this one? i though we added one in the last PR.

yati1998 replied:
in last PR it was removed as not required, maybe after rebase it's considered as a new change, I will remove it.

yati1998 commented on docs/subvolume.md:
```diff
@@ -40,16 +40,20 @@ kubectl rook-ceph subvolume ls --stale
-# Info: subvolume csi-vol-427774b4-340b-11ed-8d66-0242ac110004 deleted
-# Info: subvolume csi-vol-427774b4-340b-11ed-8d66-0242ac110004 deleted
+# Info: subvolume "csi-vol-test" deleted
+# Error: error getting omap keys ocs-storagecluster-cephfilesystem-metadata/test.subvolume:
+# (2) No such file or directory
+# . failed to run command. command terminated with exit code 1%!(EXTRA string=failed to list omapval)
```
Incase the subvolume is created from backend and it's stale, there will be no omap created for it.  In that case if we try to delete the omap, it will through this error.

subhamkrai replied:
let's document the happy case output.

subhamkrai commented on pkg/filesystem/subvolume.go:
```diff
@@ -211,7 +217,97 @@ func Delete(ctx context.Context, clientsets *k8sutil.Clientsets, OperatorNamespa
+		if snapshot {
+			logging.Error(fmt.Errorf("subvolume %s has snapshots", subvolume))
+			return false
+		} else {
+			return true
+		}
```
```suggestion
		} 
		return true
```
else is dead code here since we have return in if

yati1998 commented:
@travisn @Madhu-1 I have addressed all the comments above, please do have a look once you are free.
@Madhu-1 about adding the test for stale rados omap, we need to create sc with retain feature and a few other steps need to be followed, I am working on it to check how we can add that here in ci.

yati1998 commented on pkg/filesystem/subvolume.go:
```diff
@@ -211,7 +217,89 @@ func Delete(ctx context.Context, clientsets *k8sutil.Clientsets, OperatorNamespa
+		volname := strings.Split(ls, "|")
+		if checknfs {
+			if len(volname) > 16 {
+				pv = "csi.volume." + volname[12] + volname[14] + volname[16]
```
I am not sure, I initially tried to get the value for csi.imagename only, will see if there is a way possible

nixpanic replied:
A much cleaner way is to write to a temporary file:
```console
sh-5.1$ rados --pool ocs-storagecluster-cephblockpool getomapval csi.volume.df42d7a3-8f3a-4630-835e-ed264a088777 csi.imagename /tmp/imagename 
Writing to /tmp/imagename
sh-5.1$ grep '' /tmp/imagename
csi-vol-df42d7a3-8f3a-4630-835e-ed264a088777
```
Note that the contents of `/tmp/imagename` does not contain a `\n` at the end of the line.

You can also write to `/dev/stdout` so that you do not need to create/remove a temporary file:
```console
sh-5.1$ IMAGENAME=$(rados --pool ocs-storagecluster-cephblockpool getomapval csi.volume.df42d7a3-8f3a-4630-835e-ed264a088777 csi.imagename /dev/stdout 2> /dev/null)
sh-5.1$ echo ${IMAGENAME}
csi-vol-df42d7a3-8f3a-4630-835e-ed264a088777
```

A big _**NO**_ against the parsing of the output the way you do now, sorry!

yati1998 replied:
Yeah Yeah., I am working on it. I was bit busy testing the ci tests, will work on other comments too.

yati1998 commented on pkg/filesystem/subvolume.go:
```diff
@@ -211,7 +217,89 @@ func Delete(ctx context.Context, clientsets *k8sutil.Clientsets, OperatorNamespa
+	if err != nil {
+		if !strings.Contains(err.Error(), "No such file or directory") {
+			logging.Fatal(fmt.Errorf("failed to list omapvals: %q", err))
+		} else {
```
But what if the condition is not true and there are no omap details, in that case, it will execute the else statement. I have tested this condition.

yati1998 commented:
@subhamkrai https://github.com/rook/kubectl-rook-ceph/actions/runs/8060623451/job/22031771871?pr=237#step:22:79 any idea why the ci is failing?

subhamkrai commented on tests/github-action-helper.sh:
```diff
@@ -31,9 +31,21 @@ deploy_rook() {
+  sed -i "s|storageClassName: rook-cephfs|storageClassName: rook-cephfs-retain|g" pvc.yaml
+  kubectl create -f pvc.yaml
+  kubectl delete pvc cephfs-pvc-retain
+  "${pvname:=$(kubectl get pv | grep cephfs-pvc-retain | awk '{print $1}')}"
```
```suggestion
  PVNAME:=$(kubectl get pv | grep cephfs-pvc-retain | awk '{print $1}')
```
or 
```
:  "${PVNAME:=$(kubectl get pv | grep cephfs-pvc-retain | awk '{print $1}')}"
```

yati1998 replied:
Initially I tried with the first suggestion, but even that is failing, will go for the next.

subhamkrai commented on pkg/filesystem/subvolume.go:
```diff
@@ -211,7 +217,90 @@ func Delete(ctx context.Context, clientsets *k8sutil.Clientsets, OperatorNamespa
+			return omapDetails{}
+		}
+	}
+	if ls != "" {
+
+		// Below is how the output of listmapvals looks, and when we split it w.r.t
+		// '|' we can get the pv name at 7,9 and 11th part of the output.
+		// Below example is for cephfs, we have similar case for nfs subvolumes.
+		//csi.imagename
+		//value (44 bytes) :
+		//00000000  63 73 69 2d 76 6f 6c 2d  35 66 62 32 61 38 30 32  |csi-vol-5fb2a802|
+		//00000010  2d 38 37 31 31 2d 34 33  63 33 2d 62 39 36 35 2d  |-8711-43c3-b965-|
+		//00000020  66 66 37 65 37 39 35 62  33 35 33 65              |ff7e795b353e|
+		//0000002c
+		//csi.volname
+		//value (40 bytes) :
+		//00000000  70 76 63 2d 65 63 62 62  34 34 38 61 2d 30 35 37  |pvc-ecbb448a-057|
+		//00000010  36 2d 34 35 30 35 2d 62  32 39 35 2d 30 32 39 33  |6-4505-b295-0293|
+		//00000020  34 37 37 36 34 65 61 64                           |47764ead|
+		//0000002
+		volname := strings.Split(ls, "|")
+		if checknfs {
+			if len(volname) > 16 {
+				pv = "csi.volume." + volname[12] + volname[14] + volname[16]
+			}
+		} else {
+			if len(volname) > 11 {
+				pv = "csi.volume." + volname[7] + volname[9] + volname[11]
+			}
+		}
+
+		details = omapDetails{omapval, pv}
+	} else {
+		logging.Fatal(fmt.Errorf("omap details not found"))
+	}
```
let's keep
```
if ls == "" {
 logging.Fatal(fmt.Errorf("omap details not found"))
} else {
 happy path code
}

it's better to keep the condition first which return the code

BlaineEXE replied:
Up to subham's preference, a form that I see more often in golang code is this, which keeps code easier to read in the happy-path case:

```go
if ls == "" {
    logging.Fatal(/* blah */)
    return omapDetails{} // early return
}
// happy path on the left
```

subhamkrai replied:
so in our code
`logging.Fatal `: log error and return
`logging.Error`: just return the error

BlaineEXE replied:
Oh, I even missed the `Fatal` part. I think that means the `else` isn't necessary from your initial example.

subhamkrai replied:
right

subhamkrai commented on tests/github-action-helper.sh:
```diff
@@ -31,9 +31,24 @@ deploy_rook() {
   kubectl create -f https://raw.githubusercontent.com/rook/rook/master/deploy/examples/toolbox.yaml
+  kubectl create -f https://raw.githubusercontent.com/rook/rook/master/deploy/examples/csi/cephfs/storageclass.yaml
   kubectl create -f https://raw.githubusercontent.com/rook/rook/master/deploy/examples/csi/rbd/storageclass-test.yaml
+  curl https://raw.githubusercontent.com/rook/rook/master/deploy/examples/csi/cephfs/storageclass.yaml -o storageclass.yaml
+  sed -i "s|name: rook-cephfs|name: rook-cephfs-retain|g" storageclass.yaml
+  sed -i "s|reclaimPolicy: Delete|reclaimPolicy: Retain|g" storageclass.yaml
+  kubectl create -f storageclass.yaml
```
```suggestion
  
  curl https://raw.githubusercontent.com/rook/rook/master/deploy/examples/csi/cephfs/storageclass.yaml -o storageclass.yaml
  sed -i "s|name: rook-cephfs|name: rook-cephfs-retain|g" storageclass.yaml
  sed -i "s|reclaimPolicy: Delete|reclaimPolicy: Retain|g" storageclass.yaml
  kubectl create -f storageclass.yaml
  
```

When not' directly creating let's add a space to make it more readable

subhamkrai commented on tests/github-action-helper.sh:
```diff
@@ -31,9 +31,24 @@ deploy_rook() {
+  kubectl create -f storageclass.yaml
   kubectl create -f https://raw.githubusercontent.com/rook/rook/master/deploy/examples/csi/rbd/pvc.yaml
   kubectl create -f https://raw.githubusercontent.com/rook/rook/master/deploy/examples/csi/cephfs/pvc.yaml
+  curl https://raw.githubusercontent.com/rook/rook/master/deploy/examples/csi/cephfs/pvc.yaml -o pvc.yaml
+  sed -i "s|name: cephfs-pvc|name: cephfs-pvc-retain|g" pvc.yaml
+  sed -i "s|storageClassName: rook-cephfs|storageClassName: rook-cephfs-retain|g" pvc.yaml
+  kubectl create -f pvc.yaml
+  kubectl get pvc cephfs-pvc-retain
+  wait_for_pvc_to_be_bound_state_default
+  : "${PVNAME:=$(kubectl get pvc cephfs-pvc-retain -o=jsonpath='{.spec.volumeName}')}"
+  kubectl get pvc cephfs-pvc-retain
+  kubectl delete pvc cephfs-pvc-retain
+  kubectl delete pv "$PVNAME"
```
```suggestion

  curl https://raw.githubusercontent.com/rook/rook/master/deploy/examples/csi/cephfs/pvc.yaml -o pvc.yaml
  sed -i "s|name: cephfs-pvc|name: cephfs-pvc-retain|g" pvc.yaml
  sed -i "s|storageClassName: rook-cephfs|storageClassName: rook-cephfs-retain|g" pvc.yaml
  kubectl create -f pvc.yaml

  kubectl get pvc cephfs-pvc-retain
  wait_for_pvc_to_be_bound_state_default
  : "${PVNAME:=$(kubectl get pvc cephfs-pvc-retain -o=jsonpath='{.spec.volumeName}')}"
  kubectl get pvc cephfs-pvc-retain
  kubectl delete pvc cephfs-pvc-retain
  kubectl delete pv "$PVNAME"
```

Madhu-1 commented on .github/workflows/go-test.yaml:
```diff
@@ -82,6 +82,8 @@ jobs:
           kubectl rook-ceph subvolume ls --stale
           kubectl rook-ceph subvolume delete myfs test-subvol group-a
           kubectl rook-ceph subvolume delete myfs test-subvol-1
+          SubVol=$(kubectl rook-ceph subvolume ls --stale | awk '{print $2}' | grep csi-vol)
```
```suggestion
         subVol=$(kubectl rook-ceph subvolume ls --stale | awk '{print $2}' | grep csi-vol)
```

Madhu-1 commented on docs/subvolume.md:
````diff
@@ -39,8 +39,18 @@ kubectl rook-ceph subvolume ls --stale
+```
+
+```bash
+kubectl rook-ceph subvolume delete ocs-storagecluster-cephfilesystem csi-vol-test
````
`csi-vol-test` is not a proper format use `csi-vol-427774b4-340b-11ed-8d66-0242ac112345`

Madhu-1 commented on pkg/filesystem/subvolume.go:
```diff
@@ -211,7 +217,90 @@ func Delete(ctx context.Context, clientsets *k8sutil.Clientsets, OperatorNamespa
+	poolName := fs + "-metadata"
+	if omapdetails.omapval != "" {
+		// remove omap object.
+		exec.RunCommandInOperatorPod(ctx, clientsets, "rados", []string{"rm", omapdetails.omapval, "-p", poolName, "--namespace", svg}, OperatorNamespace, CephClusterNamespace, true)
```
`--namespace` is always `csi` we cannot pass svg value to it

Madhu-1 commented on pkg/filesystem/subvolume.go:
```diff
@@ -211,7 +217,90 @@ func Delete(ctx context.Context, clientsets *k8sutil.Clientsets, OperatorNamespa
+func deleteOmap(ctx context.Context, clientsets *k8sutil.Clientsets, OperatorNamespace, CephClusterNamespace, subVol, fs, svg string) {
+	logging.Info("Deleting the omap object and key for subvolume %q", subVol)
+	omapdetails := getOmapDetails(ctx, clientsets, OperatorNamespace, CephClusterNamespace, subVol, fs, svg)
+	poolName := fs + "-metadata"
```
AFAIK this is not the right way, we need to get the metadata pool name from the filesystem CR

Madhu-1 commented on pkg/filesystem/subvolume.go:
```diff
@@ -211,7 +217,90 @@ func Delete(ctx context.Context, clientsets *k8sutil.Clientsets, OperatorNamespa
+	}
+	if omapdetails.omapkey != "" {
+		// remove omap key.
+		exec.RunCommandInOperatorPod(ctx, clientsets, "rados", []string{"rmomapkey", "csi.volumes.default", omapdetails.omapkey, "-p", poolName, "--namespace", svg}, OperatorNamespace, CephClusterNamespace, true)
```
`exec.RunCommandInOperatorPod` can fail due to other reason we should not skip error check and consider all the cases as success isn't it?

Madhu-1 commented on pkg/filesystem/subvolume.go:
```diff
@@ -211,7 +217,90 @@ func Delete(ctx context.Context, clientsets *k8sutil.Clientsets, OperatorNamespa
+}
+
+// getOmapDetails gets the omap key and value details for a given subvolume.
+// Deletion of omap object required the rbd-image-name which is of format
```
`rbd-image-name`?

Madhu-1 commented on pkg/filesystem/subvolume.go:
```diff
@@ -211,7 +217,90 @@ func Delete(ctx context.Context, clientsets *k8sutil.Clientsets, OperatorNamespa
+	var omapval, pv string
+	poolName := fs + "-metadata"
+	replace := strings.Replace(subVol, "-", ".", 2)
+	checknfs := strings.Contains(subVol, "nfs-export")
```
i think this assumption is wrong, anyone can create subvolume for nfs without this as well :)

yati1998 replied:
Will send nfs changes in another PR. Currently, removing it from this comit to make it simpler.

Madhu-1 commented on pkg/filesystem/subvolume.go:
```diff
@@ -211,7 +217,90 @@ func Delete(ctx context.Context, clientsets *k8sutil.Clientsets, OperatorNamespa
+
+		details = omapDetails{omapval, pv}
+	} else {
+		logging.Fatal(fmt.Errorf("omap details not found"))
```
This also seems also not correct, if not found should be considered as success as it might have been already deleted.

nixpanic commented on pkg/filesystem/subvolume.go:
```diff
@@ -201,17 +208,100 @@ func unMarshaljson(list string) []fsStruct {
+	pvname, err := exec.RunCommandInOperatorPod(ctx, clientsets, "rados", cmd, OperatorNamespace, CephClusterNamespace, true)
+	if err != nil || pvname == "" {
+		if !strings.Contains(err.Error(), "No such file or directory") {
+			logging.Fatal(fmt.Errorf("failed to list omapvals: %q", err))
```
You are not listing omap values anymore, you now only get a single one. Include the missing value and error?

Is this really fatal? What if the values were cleaned up some way already?

yati1998 commented on pkg/filesystem/subvolume.go:
```diff
@@ -201,17 +208,100 @@ func unMarshaljson(list string) []fsStruct {
 	if !check {
-		exec.RunCommandInOperatorPod(ctx, clientsets, "ceph", []string{"fs", "subvolume", "rm", fs, subvol, svg, "--retain-snapshots"}, OperatorNamespace, CephClusterNamespace, true)
+		deleteOmap(ctx, clientsets, OperatorNamespace, CephClusterNamespace, subvol, fs)
+		_, err := exec.RunCommandInOperatorPod(ctx, clientsets, "ceph", []string{"fs", "subvolume", "rm", fs, subvol, svg, "--retain-snapshots"}, OperatorNamespace, CephClusterNamespace, true)
```
if snapshots are present we can't simply delete the subvolumes, we first need to delete the snapshots and then the subvolume. And I am not very sure above deleting the snapshots from admin side. Hence, best way to delete the subvolume would be with retain-snapshots.
cc @nixpanic  @Madhu-1 what do you think?

nixpanic replied:
Ceph-CSI uses a reference counter in OMAP, I think... Only when the reference counter is 0, the subvolume/snapshot should be deleted. Maybe that is only for snapshots, not sure it is used for (plain) subvolumes too.

Madhu-1 replied:
Yes snapshots are meant for backup , even if the subvolume is stale for some reason, the user can restore from the snapshots later

yati1998 commented on pkg/filesystem/subvolume.go:
```diff
@@ -201,17 +208,100 @@ func unMarshaljson(list string) []fsStruct {
+		_, err := exec.RunCommandInOperatorPod(ctx, clientsets, "ceph", []string{"fs", "subvolume", "rm", fs, subvol, svg, "--retain-snapshots"}, OperatorNamespace, CephClusterNamespace, true)
+		if err != nil {
+			logging.Error(err, "failed to delete subvolume of %q/%q", fs, subvol)
+		}
```
ahh, missed this.

Madhu-1 commented on pkg/filesystem/subvolume.go:
```diff
@@ -201,17 +208,98 @@ func unMarshaljson(list string) []fsStruct {
 	if !check {
-		exec.RunCommandInOperatorPod(ctx, clientsets, "ceph", []string{"fs", "subvolume", "rm", fs, subvol, svg, "--retain-snapshots"}, OperatorNamespace, CephClusterNamespace, true)
-		logging.Info("subvolume %q deleted", subvol)
+		deleteOmap(ctx, clientsets, OperatorNamespace, CephClusterNamespace, subvol, fs)
+		_, err := exec.RunCommandInOperatorPod(ctx, clientsets, "ceph", []string{"fs", "subvolume", "rm", fs, subvol, svg, "--retain-snapshots"}, OperatorNamespace, CephClusterNamespace, true)
```
first delete the subvolume and than delete the rados object?

yati1998 replied:
Incase the subvolume is deleted and rados omap deletion gets stuck, we will loose the info to delete the rados omap too. So from my knowledge and testing I would consider deleting the omap first and then subvolume.

Madhu-1 commented on pkg/filesystem/subvolume.go:
```diff
@@ -201,17 +208,98 @@ func unMarshaljson(list string) []fsStruct {
+		deleteOmap(ctx, clientsets, OperatorNamespace, CephClusterNamespace, subvol, fs)
+		_, err := exec.RunCommandInOperatorPod(ctx, clientsets, "ceph", []string{"fs", "subvolume", "rm", fs, subvol, svg, "--retain-snapshots"}, OperatorNamespace, CephClusterNamespace, true)
+		if err != nil {
+			logging.Error(err, "failed to delete subvolume of %q/%q", fs, subvol)
```
log should be fetal isnt it?
log in format of `fsName/groupName/subVolName`

Madhu-1 commented on pkg/filesystem/subvolume.go:
```diff
@@ -201,17 +208,98 @@ func unMarshaljson(list string) []fsStruct {
+		if err != nil {
+			logging.Error(err, "failed to delete subvolume of %q/%q", fs, subvol)
+		} else {
+			logging.Info("subvolume %q deleted", subvol)
```
follow same format for all the cases

Madhu-1 commented on pkg/filesystem/subvolume.go:
```diff
@@ -201,17 +208,98 @@ func unMarshaljson(list string) []fsStruct {
+	// getFilesystem gets the filesystem
+	fsstruct, err := getFileSystem(ctx, clientsets, OperatorNamespace, CephClusterNamespace)
+	if err != nil {
+		logging.Error(err, "failed to get filesystem")
```
lets not log in the internal functions, log it from the caller

Madhu-1 commented on pkg/filesystem/subvolume.go:
```diff
@@ -201,17 +208,98 @@ func unMarshaljson(list string) []fsStruct {
 	}
 }
+
+func getPoolName(ctx context.Context, clientsets *k8sutil.Clientsets, OperatorNamespace, CephClusterNamespace, fs string) (string, error) {
```
it should be get metadatPool Name not just pool name isn't it?

Madhu-1 commented on pkg/filesystem/subvolume.go:
```diff
@@ -201,17 +208,98 @@ func unMarshaljson(list string) []fsStruct {
+		// remove omap object.
+		_, err := exec.RunCommandInOperatorPod(ctx, clientsets, "rados", []string{"rm", omapdetails.omapval, "-p", poolName, "--namespace", "csi"}, OperatorNamespace, CephClusterNamespace, true)
+		if err != nil {
+			logging.Error(err, "failed to remove omap object for subvolume %q", subVol)
```
should be logged as fatal or return from here or else logging.Info at line 261 will also gets logged

Madhu-1 commented on pkg/filesystem/subvolume.go:
```diff
@@ -201,17 +208,98 @@ func unMarshaljson(list string) []fsStruct {
+		// remove omap key.
+		_, err := exec.RunCommandInOperatorPod(ctx, clientsets, "rados", []string{"rmomapkey", "csi.volumes.default", omapdetails.omapkey, "-p", poolName, "--namespace", "csi"}, OperatorNamespace, CephClusterNamespace, true)
+		if err != nil {
+			logging.Error(err, "failed to remove omap key for subvolume %q", subVol)
+		}
+		logging.Info("omap key:%q deleted", omapdetails.omapkey)
```
same here as well

Madhu-1 commented on pkg/filesystem/subvolume.go:
```diff
@@ -201,17 +208,98 @@ func unMarshaljson(list string) []fsStruct {
+	if err != nil || poolName == "" {
+		logging.Fatal(fmt.Errorf("pool name not found: %q", err))
+	}
+	// omapval is of format csi.volume.427774b4-340b-11ed-8d66-0242ac110005
+	// which is similar to volume name csi-vol-427774b4-340b-11ed-8d66-0242ac110005
+	// hence, replacing '-' to '.' to get the omapval
```
instead of replace why not remove `csi-vol-` and use it? and this might not work if someone is using https://github.com/ceph/ceph-csi/blob/c90f7ed7776e16c07a07c62106b7643a2819279e/examples/cephfs/storageclass.yaml#L48 volumePrefix

Madhu-1 commented on pkg/filesystem/subvolume.go:
```diff
@@ -201,17 +208,98 @@ func unMarshaljson(list string) []fsStruct {
+// omapkey is the pv name which is extracted the omap object.
+func getOmapDetails(ctx context.Context, clientsets *k8sutil.Clientsets, OperatorNamespace, CephClusterNamespace, subVol, fs string) omapDetails {
+
+	var details omapDetails
+	var omapval, omapkey string
```
we dont need var declaration right? we can declare these variables where its required.

yati1998 commented:
@travisn @subhamkrai @sp98 can anyone please check if there is some issue with rbd command, ci is getting stuck there.
Also @travisn have addressed your comments.

sp98 commented:
> @travisn @subhamkrai @sp98 can anyone please check if there is some issue with rbd command, ci is getting stuck there. Also @travisn have addressed your comments.

in the tmate session the same command is working. Only difference is that it doesn't return anything. But I don't think that should cause the CI to stuck!

``` 
runner@fv-az527-340:~/work/kubectl-rook-ceph/kubectl-rook-ceph$ kubectl rook-ceph  rbd ls replicapool
Warning: rook version 'rook: v1.13.0-alpha.0.389.gc409a6ae4' is running a pre-release version of Rook.

runner@fv-az527-340:~/work/kubectl-rook-ceph/kubectl-rook-ceph$ 
```

yati1998 commented on pkg/filesystem/subvolume.go:
```diff
@@ -44,6 +46,11 @@ const (
 	staleWithSnapshot = "stale-with-snapshot"
 )
 
+type omapDetails struct {
```
forgot to remove, thanks.

yati1998 commented on pkg/filesystem/subvolume.go:
```diff
@@ -201,17 +208,106 @@ func unMarshaljson(list string) []fsStruct {
+}
+
+// deleteOmap deletes omap object and key for the given subvolume.
+func deleteOmap(ctx context.Context, clientsets *k8sutil.Clientsets, OperatorNamespace, CephClusterNamespace, subVol, fs string) {
```
we are not deleting omap from subvolume. In csi on creation of subvolume, it is stored in omap for record. I don't think this could be a suitable name.

yati1998 replied:
got it, will update it

subhamkrai commented:
```
persistentvolumeclaim "cephfs-pvc-retain" deleted
+ kubectl delete pv ''
error: resource name may not be empty
```

Pv is deleted or not present

yati1998 commented:
```runner@fv-az1154-938:~/work/kubectl-rook-ceph/kubectl-rook-ceph$ oc get pvc
NAME                 STATUS    VOLUME   CAPACITY   ACCESS MODES   STORAGECLASS      AGE
cephfs-pvc           Pending                                      rook-cephfs       18m
cephfs-pvc-retain1   Pending                                      rook-cephfs       108s
rbd-pvc              Pending                                      rook-ceph-block   18m
```

I see none of the pvcs are getting created and the error being `volumeID already exists`. Seems we need to have some cleanup steps before we create the resources . Anyone has any idea why is it so?
cc @subhamkrai @travisn @Madhu-1

yati1998 commented:
> ```
> persistentvolumeclaim "cephfs-pvc-retain" deleted
> + kubectl delete pv ''
> error: resource name may not be empty
> ```
> 
> Pv is deleted or not present

The pv is not present as the pvc is not in the bound state. And the reason for it in mentioned in my last comment.

nixpanic commented on pkg/filesystem/subvolume_test.go:
```diff
@@ -0,0 +1,43 @@
+	tests := []struct {
+		name string
+		want string
+	}{
```
Maybe consider adding negative tests as well? An empty string, a string with only no, one or two `-` in there probably are problematic and cause issues. It may be unlikely that these get passed to the function, but you should be prepared for weird issues in a cleanup process, some values may be unexpectedly missing or somehow got corrupted.

yati1998 commented:
@Madhu-1 waiting for your review here. Please feel free to add your comments.

Madhu-1 commented on pkg/filesystem/subvolume.go:
```diff
@@ -201,17 +203,109 @@ func unMarshaljson(list string) []fsStruct {
+}
+
+func getMetadataPoolName(ctx context.Context, clientsets *k8sutil.Clientsets, OperatorNamespace, CephClusterNamespace, fs string) (string, error) {
+	// getFilesystem gets the filesystem
```
can we remove this comment?

Madhu-1 commented on pkg/filesystem/subvolume.go:
```diff
@@ -201,17 +203,109 @@ func unMarshaljson(list string) []fsStruct {
+		return "", err
+	}
+
+	// this iterates over the filesystems and subvolumegroup to get the list of subvolumes that exist
```
this comments also need to be removed, its talking about group,subvolumes etc.

Madhu-1 commented on pkg/filesystem/subvolume.go:
```diff
@@ -201,17 +203,109 @@ func unMarshaljson(list string) []fsStruct {
+			return pool.MetadataName, nil
+		}
+	}
+	return "", fmt.Errorf("pool not found")
```
return something like `fmt.Errorf("metadataPool not found for %q filesystem",fs)`

Madhu-1 commented on pkg/filesystem/subvolume.go:
```diff
@@ -201,17 +203,109 @@ func unMarshaljson(list string) []fsStruct {
+	omapkey := getOmapKey(ctx, clientsets, OperatorNamespace, CephClusterNamespace, subVol, fs)
+	omapval := getOmapVal(subVol)
+	poolName, err := getMetadataPoolName(ctx, clientsets, OperatorNamespace, CephClusterNamespace, fs)
+	if err != nil || poolName == "" {
+		logging.Fatal(fmt.Errorf("pool name not found: %q", err))
```
we dont return err as nil and empty PoolName isn't it?

yati1998 replied:
yess

Madhu-1 commented on pkg/filesystem/subvolume.go:
```diff
@@ -201,17 +203,109 @@ func unMarshaljson(list string) []fsStruct {
+func getOmapKey(ctx context.Context, clientsets *k8sutil.Clientsets, OperatorNamespace, CephClusterNamespace, subVol, fs string) string {
+
+	poolName, err := getMetadataPoolName(ctx, clientsets, OperatorNamespace, CephClusterNamespace, fs)
+	if err != nil || poolName == "" {
+		logging.Fatal(fmt.Errorf("pool name not found: %q", err))
```
same here as well

yati1998 replied:
yes, it's not required.

Madhu-1 commented on pkg/filesystem/subvolume.go:
```diff
@@ -201,17 +203,109 @@ func unMarshaljson(list string) []fsStruct {
+	cmd := []string{"getomapval", omapval, "csi.volname", "-p", poolName, "--namespace", "csi", "/dev/stdout"}
+	pvname, err := exec.RunCommandInOperatorPod(ctx, clientsets, "rados", cmd, OperatorNamespace, CephClusterNamespace, true)
+	if err != nil || pvname == "" {
+		logging.Info("No pvname found for subvolume %s: %s", subVol, err)
```
```suggestion
logging.Info("No PV found for subvolume %s: %s", subVol, err)
```

Madhu-1 reviewed (commented):
LGTM,
small nits

subhamkrai commented on pkg/filesystem/subvolume.go:
```diff
@@ -201,17 +203,107 @@ func unMarshaljson(list string) []fsStruct {
+		_, err := exec.RunCommandInOperatorPod(ctx, clientsets, "ceph", []string{"fs", "subvolume", "rm", fs, subvol, svg, "--retain-snapshots"}, OperatorNamespace, CephClusterNamespace, true)
+		if err != nil {
+			logging.Fatal(err, "failed to delete subvolume of %s/%s/%s", fs, svg, subvol)
+		} else {
+			logging.Info("subvolume %s/%s/%s deleted", fs, svg, subvol)
+		}
```
```suggestion
		} 
			logging.Info("subvolume %s/%s/%s deleted", fs, svg, subvol)
		
```
this else is not required as `logging.Fatal` return and exit the code

subhamkrai commented on pkg/filesystem/subvolume.go:
```diff
@@ -201,17 +203,107 @@ func unMarshaljson(list string) []fsStruct {
+		// remove omap object.
+		_, err := exec.RunCommandInOperatorPod(ctx, clientsets, "rados", []string{"rm", omapval, "-p", poolName, "--namespace", "csi"}, OperatorNamespace, CephClusterNamespace, true)
+		if err != nil {
+			logging.Error(err, "failed to remove omap object for subvolume %q", subVol)
```
```suggestion
			logging.Error(err, "failed to remove omap object for subvolume %q", subVol)
```
`logging.Error` will only log and not return and exit and since this method doesn't return anything do you think we should use `logging.Fatal`?

subhamkrai commented on pkg/filesystem/subvolume.go:
```diff
@@ -201,17 +203,107 @@ func unMarshaljson(list string) []fsStruct {
+		// remove omap key.
+		_, err := exec.RunCommandInOperatorPod(ctx, clientsets, "rados", []string{"rmomapkey", "csi.volumes.default", omapkey, "-p", poolName, "--namespace", "csi"}, OperatorNamespace, CephClusterNamespace, true)
+		if err != nil {
+			logging.Error(err, "failed to remove omap key for subvolume %q", subVol)
```
```suggestion
			logging.Error(err, "failed to remove omap key for subvolume %q", subVol)
```
same here

subhamkrai commented on tests/github-action-helper.sh:
```diff
@@ -109,6 +140,15 @@ deploy_csi_driver_custom_ns() {
+      echo "waiting for the pvc to be in bound state"
+      sleep 1
+    done
+EOF
+}
```
```suggestion
EOF
    timeout_command_exit_code
}
```

add above function call to exit if timeout reached

subhamkrai reviewed (changes requested):
Also, since this clean cephfs OMAP please update the commit title and PR title to mention cephFS OMAP

subhamkrai commented on docs/subvolume.md:
````diff
@@ -17,30 +17,41 @@ The subvolume command will require the following sub commands:
+myfs csi-vol-427774b4-340b-11ed-8d66-0242ac110005 csi in-use
+myfs csi-vol-427774b4-340b-11ed-8d66-0242ac110007 csi stale
+myfs csi-vol-427774b4-340b-11ed-8d66-0242ac110007 csi stale-with-snapshot
 
 ```
````
```suggestion
```
```

remove extra line

subhamkrai commented on docs/subvolume.md:
````diff
@@ -17,30 +17,41 @@ The subvolume command will require the following sub commands:
-# ocs-storagecluster-cephfilesystem csi-vol-427774b4-340b-11ed-8d66-0242ac110005 csi stale-with-snapshot
+Filesystem  Subvolume  SubvolumeGroup state
+myfs csi-vol-427774b4-340b-11ed-8d66-0242ac110004 csi stale
+myfs csi-vol-427774b4-340b-11ed-8d66-0242ac110005 csi stale-with-snapshot
 
 ```
````
remove extra line

subhamkrai commented on docs/subvolume.md:
````diff
@@ -17,30 +17,41 @@ The subvolume command will require the following sub commands:
+$ kubectl rook-ceph subvolume delete myfs csi-vol-427774b4-340b-11ed-8d66-0242ac110005
+
+Info: No omapvals found for subvolume csi-vol-427774b4-340b-11ed-8d66-0242ac110005
+Info: subvolume "csi-vol-427774b4-340b-11ed-8d66-0242ac110005" deleted
 
 ```
````
```suggestion
Info: subvolume "csi-vol-427774b4-340b-11ed-8d66-0242ac110005" deleted
```
```

remove extra line

subhamkrai commented on pkg/filesystem/subvolume.go:
```diff
@@ -201,17 +203,107 @@ func unMarshaljson(list string) []fsStruct {
+		_, err := exec.RunCommandInOperatorPod(ctx, clientsets, "rados", []string{"rmomapkey", "csi.volumes.default", omapkey, "-p", poolName, "--namespace", "csi"}, OperatorNamespace, CephClusterNamespace, true)
+		if err != nil {
+			logging.Fatal(err, "failed to remove omap key for subvolume %q", subVol)
+		} else {
+			logging.Info("omap key:%q deleted", omapkey)
+		}
```
```suggestion
		} 
			logging.Info("omap key:%q deleted", omapkey)
		
```

we have `logging.Fatal` so no need to use else here

subhamkrai commented on pkg/filesystem/subvolume.go:
```diff
@@ -201,17 +203,107 @@ func unMarshaljson(list string) []fsStruct {
+		_, err := exec.RunCommandInOperatorPod(ctx, clientsets, "rados", []string{"rm", omapval, "-p", poolName, "--namespace", "csi"}, OperatorNamespace, CephClusterNamespace, true)
+		if err != nil {
+			logging.Fatal(err, "failed to remove omap object for subvolume %q", subVol)
+		} else {
+			logging.Info("omap object:%q deleted", omapval)
+		}
```
```suggestion
		} else {
			logging.Info("omap object:%q deleted", omapval)
		}
```
same
